src/lib/misc/roughtime/roughtime.cpp

```diff
@@ -0,0 +1,370 @@
+    tags[label] = std::vector<uint8_t>(buf+start, buf+end);
+    start = end;
+    }
+  return tags;
+  }
+
+template<typename T>
+T get(const std::map<std::string, std::vector<uint8_t>>& map, const std::string& label)
+  {
+    const auto& tag = map.find(label);
+    if(tag == map.end())
+      throw Invalid_Argument("tag not found");
+    if(tag->second.size() != sizeof(T))
+      throw Invalid_Argument("tag content wrong size");
+    return *(const T*)tag->second.data();
+  }
+
+const std::vector<uint8_t>& get_v(const std::map<std::string, std::vector<uint8_t>>& map, const std::string& label)
+  {
+    const auto& tag = map.find(label);
+    if(tag == map.end())
+      throw Invalid_Argument("tag not found");
+    return tag->second;
+  }
+
+bool verify_signature(const std::array<uint8_t,32>& pk, const std::vector<uint8_t>& payload, const std::array<uint8_t,64>& signature)
+  {
+  const char context[] = "RoughTime v1 response signature";
+  Ed25519_PublicKey key(std::vector<uint8_t>(pk.data(), pk.data()+pk.size()));
+  PK_Verifier verifier(key, "Pure");
+  verifier.update((const uint8_t*)context, sizeof(context)); //add context including \0
+  verifier.update(payload);
+  return verifier.check_signature(std::vector<uint8_t>(signature.data(), signature.data()+signature.size()));
+  }
+
+std::array<uint8_t,64> hashLeaf(const std::array<uint8_t,64>& leaf)
+  {
+  std::array<uint8_t,64> ret;
+  std::unique_ptr<HashFunction> hash(HashFunction::create("SHA-512"));
+  hash->update(0);
+  hash->update(leaf.data(), leaf.size());
+  hash->final(ret.data());
+  return ret;
+  }
+
+std::array<uint8_t,64> hashNode(const std::array<uint8_t,64>& left, const std::array<uint8_t,64>& right)
+  {
+  std::array<uint8_t,64> ret;
+  std::unique_ptr<HashFunction> hash(HashFunction::create("SHA-512"));
+  hash->update(1);
+  hash->update(left.data(), left.size());
+  hash->update(right.data(), right.size());
+  hash->final(ret.data());
+  return ret;
+  }
+
+template<size_t N>
+std::array<uint8_t, N> vector_to_array(std::vector<uint8_t> vec)
+  {
+  if(vec.size() != N)
+    throw Invalid_Argument("invalid vector size");
+  return *(std::array<uint8_t, N>*)vec.data();
+  }
+}
+
+namespace Roughtime {
+
+std::array<uint8_t, request_min_size> encode_request(const std::array<uint8_t, 64>& nonce)
+  {
+  std::array<uint8_t, request_min_size> buf = {2,0,0,0,64,0,0,0,'N','O','N','C','P','A','D',0xff};
+  std::memcpy(buf.data() + 16, nonce.data(), nonce.size());
+  std::memset(buf.data() + 16 + nonce.size(), 0, buf.size() - 16 - nonce.size());
+  return buf;
+  }
+
+Response Response::from_bits(const std::vector<uint8_t>& response,
+                             const std::array<uint8_t, 64>& nonce)
+  {
+  const auto response_v = split(response);
+  const auto cert = split(get_v(response_v,"CERT"));
+  const auto cert_dele = get<std::array<uint8_t,72>>(cert,"DELE");
+  const auto cert_sig =  get<std::array<uint8_t,64>>(cert,"SIG");
```

Comment:
Is this ordering in the response guaranteed by the protocol?

Reply:
I don't see I make any assumptions on order. I construct a map with the tags, and then look them up, throwing if they don't exist of if size is not exactly the one expected.

Reply:
Right sorry